sentinel-core/src/main/java/com/alibaba/csp/sentinel/metric/extension/AdvancedMetricExtension.java

```diff
@@ -0,0 +1,97 @@
+package com.alibaba.csp.sentinel.metric.extension;
+
+import com.alibaba.csp.sentinel.EntryType;
+import com.alibaba.csp.sentinel.slots.block.BlockException;
+
+/**
+ * Advanced {@link MetricExtension} extending input parameters of each metric
+ * collection method with the name of {@link EntryType}.
+ * 
+ * @author bill_yip
+ * @since 1.8.0
+ */
+public interface AdvancedMetricExtension extends MetricExtension {
+	/**
+	 * Add current pass count of the resource name.
+	 *
+	 * @param n         count to add
+	 * @param resource  resource name
+	 * @param entryType {@link EntryType} name, [IN] as provider, [OUT] as consumer.
+	 * @param args      additional arguments of the resource, eg. if the resource is
+	 *                  a method name, the args will be the parameters of the
+	 *                  method.
+	 */
+	void addPass(String resource, String entryType, int n, Object... args);
```

Comment:
Maybe we could use `EntryType` type directly?

Reply:
I had thought about the two points you mentioned above.

The first:
Take my case for example, the original MetricExtension had been introduced into our projects.I think that If someone don't need the extra tags of metrics, he/she should not be affected by using the new version.So I added  the AdvancedMetricExtension extending the original one is for backwards compatibility.

The second:
Take MetricExtension for example, It is always being implemented in the hosted application or a common utils lib of a company. Especially when implementing it in a common utils lib, I think it is not so good to introduce sentinel-core just for the class EntryType.

Reply:
> I had thought about the two points you mentioned above.
> 
> The first:
> Take my case for example, the original **MetricExtension** had been introduced into our projects.I think that If someone don't need the extra tags of metrics, he/she should not be affected by using the new version.So I added the **AdvancedMetricExtension** extending the original one is for backwards compatibility.

Yeah, you're right. Backwards compatibility is important. But providing extra `AdvancedMetricExtension` for just extra EntryType info seems confusing (a kind of temporary solution)... It's okay for me to merge it first and improve it in later versions.

> The second:
> Take **MetricExtension** for example, It is always being implemented in the hosted application or a common utils lib of a company. Especially when implementing it in a common utils lib, I think it is not so good to introduce sentinel-core just for the class EntryType.

The `MetricExtension` itself is part of sentinel-core so I don't think it's a problem. You could mark the dependency "provided" in the common libraries.